# MenDelSIM

### Mendelian Disease Simulator 

## Introduction 

Human mendelian disease are characterized as genetic disorders caused by changes in a single loci or gene. These genetic disorders follow different inheritance patterns depending on the type of alteration in the loci. Many rare genetic diseases are mendelian in nature some examples include: cystic fibrosis, hemophilia, and color blindness. 

The MenDelSIM package allows for the simulation of new or known variants in a single gene representing a possible mendelian disease case aswell as simulation of full family profiles following a specified inheritance pattern. 

![](https://github.com/tamario/variant_simulation_framework/blob/master/misc/overview.png =250x)

![](https://github.com/tamario/variant_simulation_framework/blob/master/misc/inheritance_1.png =250x)

![](https://github.com/tamario/variant_simulation_framework/blob/master/misc/inheritance_1.png =250x)

## Directory Structure 

```
.
├── environment.yml
├── LICENCE
├── misc
├── README.md
├── sample_simulations
└── simulator
    ├── __init__.py
    ├── GUD
    ├── tests
    ├── scripts
    │   ├── cli.py
    │   ├── config_gen.py
    │   ├── family_vcf_gen.py
    │   └── __init__.py
    ├── clinvar.py
    ├── indel.py
    ├── short_tandem_repeat.py
    ├── single_nucleotide_variant.py
    ├── transcript.py
    ├── variant.py
    └── variants.py
```

## Dependancies 

- conda 
- linux/mac

## Modules 

1. -**Config Generator** 

Under `simulator/scripts/config_gen.py` this module is used to to generate config files. This module can be called using the following command `python -m simulator.scripts.config_gen`. 

2. **Variant Simulator**

Under `simulator/scripts/cli.py` this module is used to to generate vcf files from config files. This module can be called by the following command:

```
python -m simulator.scripts.cli -c <congif.json> -o <output>
## <config.json> is the config file, by default it will look for config.json .
## <output> is the name of the output, by default it will name the output: output.vcf .
```

3. **Family Generator** 

Under `simulator/scripts/family_vcf_gen.py` this module is used to to generate vcf files for a single family from the proband's vcf and the family PED structure

```
python -m simulator.scripts.family_vcf_gen -p <proband.vcf> -f <family.ped>
## <proband.vcf> is VCF of the proband as generated by the simulator.
## <family.ped> is the PED file that represents the family structure of the proband. 
```

Assumptions and Requirements: 

- the PED file is a TAB delimited file with a header following this structure:

```
#FamilyID	SampleID	Father	Mother	Sex	Phenotype
```

- The sex can must be either 1=male or 2=female
- The Phenotype can be one of the following: 0=unknown, 1=unaffected,  2=affected, 3=carrier/
- The members of the PED must include the proband and one or more of the following: proband's sibling, proband's mother, proband's father.
- The PED file must contain an entry with SampleID="Proband" representing the proband. 
- Parents will have up to 1 variant each 
- Siblings can have up to 2 variants each 

## Quickstart 

Follow these instructions to set up the package: 

```
git clone https://github.com/tamario/variant_simulation_framework.git
cd variant_simulation_framework
conda env create -f environment.yml
source activate var_sim
cd simulator/GUD
python setup.py install
```

## Running 

The simulator currently has two major points of entry, the configuration generator and the simulator. The standard workflow to run would be this 

1. Run configuration generator to generate config file that follows the EBNF. Instructions on what to enter are in the config_generator 

```
source activate var_sim
python -m simulator.scripts.config_gen
```

2. Run simulator to get vcfs

```
source activate var_sim
python -m simulator.scripts.cli -c <congig file name> -o <output file name>
```

3. Run family generator to get family vcfs

```
source activate var_sim
python -m simulator.scripts.family_vcf_gen -p <proband.vcf> -f <family.ped>
```



## Examples

Some example simulations can be found under the `sample_simulations` directory with the complete simulations for several inheritance patterns. One running example can be found below using the autosomal recessive homozygous  simulation. 

```
## directory
sample_simulations/
└── AR_homozygous
    ├── AR_homozygous.json
    ├── AR_homozygous.ped
    ├── AR_homozygous.vcf
    ├── Brother.vcf
    ├── Dad.vcf
    ├── Mom.vcf
    ├── Proband.vcf
    ├── Sister1.vcf
    └── Sister2.vcf

## running the following will generate the family VCF files
source activate var_sim

python -m simulator.scripts.cli -c sample_simulations/AR_homozygous/AR_homozygous.json -o sample_simulations/AR_homozygous/AR_homozygous.vcf

python -m simulator.scripts.family_vcf_gen -p sample_simulations/AR_homozygous/AR_homozygous.vcf -f sample_simulations/AR_homozygous/AR_homozygous.ped
```

## EBNF 

```
CONFIG ::= '{'GENE_UID','GENE_NAME', 'VAR1', 'VAR2'}'

GENE_UID ::= number
GENE_NAME ::= inputstring
VAR1 ::= VARIANT_TEMPLATE
VAR2 ::= 'NONE' | VARIANT_TEMPLATE

VARIANT_TEMPLATE ::= '{'TYPE', 'REGION', 'LOCATION', 'IMPACT', 'ZYGOSITY'}'

TYPE ::= 'SNV' | 'INDEL' | 'STR' | 'ClinVar'
REGION ::= 'CODING' | 'UTR' | 'INTRONIC'
LOCATION ::= number | 'ANY' | 'NONE'
ZYGOSITY ::= 'HETEROZYGOUS' | 'HOMOZYGOUS' | 'HEMIZYGOUS'
STR ::= {chrom, start, end} 
IMPACT ::=  
SNV_IMPACT_CODING := 'A' | 'T' | 'G' | 'C' | 'ANY' | ''MISSENSE | 'NONSENSE' | 'SILENT'
SNV_IMPACT_NONCODING := 'A' | 'T' | 'G' | 'C' | 'ANY'
INDEL_IMPACT ::= number
STR_IMPACT ::= number ## negative number is deletion of that size*motif size same logic for insertion 
ClinVar_IMPACT ::= number ## clinvarID 
```
